### PR TITLE
Only emit each warning once per source location

### PR DESCRIPTION
**Pull request by nex3**
See https://github.com/sass/sass-spec/pull/1656

### CODE REVIEW

File: lib/src/visitor/async_evaluate.dart [hunk 1]
```diff
@@ -1987,9 +1993,13 @@ class _EvaluateVisitor
...
         await _addExceptionSpanAsync(node, () => node.expression.accept(this));
-    _logger.warn(
-        value is SassString ? value.text : _serialize(value, node.expression),
-        trace: _stackTrace(node.span));
+    var message =
+        value is SassString ? value.text : _serialize(value, node.expression);
+
+    // Don't use [_warn] because we don't want to pass the span to the logger.
+    if (_warningsEmitted.add(Tuple2(message, node.span))) {
+      _warn(message, node.span);
```
Comment (Awjin) [inline lib/src/visitor/async_evaluate.dart:2001]: Should this be `_logger.warn()`?

File: lib/src/visitor/async_evaluate.dart [hunk 2]
```diff
@@ -1987,9 +1993,13 @@ class _EvaluateVisitor
...
     var value =
         await _addExceptionSpanAsync(node, () => node.expression.accept(this));
-    _logger.warn(
-        value is SassString ? value.text : _serialize(value, node.expression),
-        trace: _stackTrace(node.span));
+    var message =
+        value is SassString ? value.text : _serialize(value, node.expression);
+
+    // Don't use [_warn] because we don't want to pass the span to the logger.
+    if (_warningsEmitted.add(Tuple2(message, node.span))) {
```
Comment (stof) [inline lib/src/visitor/async_evaluate.dart:2000]: should `@warn` be deduplicated too ?

For warnings triggered in dart code, the span is the place that you actually need to change to resolve the warning (at least in most cases).
For userland functions triggering a warning with `@warn` when being misused, the span used here would be the location of the `@warn` rule inside the function, not the location of the caller (which is what where the change would have to be done).

File: lib/src/visitor/async_evaluate.dart (hunk 1 above)
Comment (stof) [reply lib/src/visitor/async_evaluate.dart:2001]: the comment above makes me thing it should

File: lib/src/visitor/async_evaluate.dart (hunk 2 above)
Comment (nex3) [reply lib/src/visitor/async_evaluate.dart:2000]: Hmm, that's a good point. We don't necessarily have a good understanding of where the boundary is between "user code" and "library code" such that we can reliably tell which stack frame has the span that should actually be used for uniqueness here. I'll drop the `@warn` de-duplication.

**Review by Awjin — APPROVED** [review]

**Pull request merged**